src/durableorchestrationcontext.ts

```diff
@@ -0,0 +1,177 @@
+import { ITaskMethods, RetryOptions, Task, TimerTask } from "./classes";
+
+/**
+ * Parameter data for orchestration bindings that can be used to schedule
+ * function-based activities.
+ */
+export class DurableOrchestrationContext {
+    /**
+     * The ID of the current orchestration instance.
+     *
+     * The instance ID is generated and fixed when the orchestrator function is
+     * scheduled. It can be either auto-generated, in which case it is
+     * formatted as a GUID, or it can be user-specified with any format.
+     */
+    public readonly instanceId: string;
+
+    /**
+     * Gets a value indicating whether the orchestrator function is currently
+     * replaying itself.
+     *
+     * This property is useful when there is logic that needs to run only when
+     * the orchestrator function is _not_ replaying. For example, certain types
+     * of application logging may become too noisy when duplicated as part of
+     * orchestrator function replay. The orchestrator code could check to see
+     * whether the function is being replayed and then issue the log statements
+     * when this value is `false`.
+     */
+    public readonly isReplaying: boolean;
+
+    /**
+     * The ID of the parent orchestration of the current sub-orchestration
+     * instance. The value will be available only in sub-orchestrations.
+     *
+     * The parent instance ID is generated and fixed when the parent
+     * orchestrator function is scheduled. It can be either auto-generated, in
+     * which case it is formatted as a GUID, or it can be user-specified with
+     * any format.
+     */
+    public readonly parentInstanceId: string;
+
+    /**
+     * Gets the current date/time in a way that is safe for use by orchestrator
+     * functions.
+     *
+     * This date/time value is derived from the orchestration history. It
+     * always returns the same value at specific points in the orchestrator
+     * function code, making it deterministic and safe for replay.
+     */
+    public currentUtcDateTime: Date;
+
+    /**
+     *
+     */
+    public Task: ITaskMethods;
+
+    /**
+     * Schedules an activity function named `name` for execution.
+     *
+     * @param name The name of the activity function to call.
+     * @param input The JSON-serializable input to pass to the activity
+     * function.
+     * @returns A Durable Task that completes when the called activity
+     * function completes or fails.
+     */
+    public callActivity(name: string, input?: unknown): Task {
+        throw new Error("Not implemented");
```

Comment:
Changed; agreed, that is more descriptive. The explanation for this is a weird one:

The `DurableOrchestrationContext` type is used to specify the shape of the [`df` property on `IFunctionContext`](https://github.com/Azure/azure-functions-durable-js/blob/3151cfd85b67e1a29521788ee6d24423fe471458/src/ifunctioncontext.ts#L7), which was previously `any` (as in -thing, which is not good practice.) It exists solely to allow for more explicit type definition within the code and be a place to hang documentation from. (Originally it was an interface, but typedoc doesn't render function parameter information well on interfaces or abstract classes.)

Its constructor is never called and never should be, hence the exception-throwing (and since the type isn't exported, would take some digging to do so.) Instead, the `df.orchestrator` wrapper function attaches an [anonymous object shaped like `DurableOrchestrationContext`](https://github.com/Azure/azure-functions-durable-js/blob/3151cfd85b67e1a29521788ee6d24423fe471458/src/orchestrator.ts#L38) whose method properties are bound to the actual implementations to the `IFunctionContext` at runtime. The TypeScript compiler is satisfied and the documentation generates nicely.